core/store/src/trie/split.rs

```diff
@@ -334,50 +369,57 @@ where
         // into the first child.
         if self.nibbles.len() % 2 != 0 {
-            self.force_next_step();
-            if let Some(current_split) = self.best_split_at_current_path() {
+            self.force_next_step()?;
+            if let Some(current_split) = self.best_split_at_current_path()? {
                 if current_split.mem_diff() < best_split.mem_diff() {
                     best_split = current_split;
                 }
             }
         }
 
-        debug_assert!(best_split.mem_diff() < u64::MAX);
-        best_split
+        if best_split.mem_diff() < u64::MAX {
+            Ok(best_split)
+        } else {
+            Err(SplitError::SplitNotFound)
+        }
     }
 
     /// Find the next step `(nibble, child_mem_usage, left_mem_usage)`.
     ///     * `nibble` – next nibble on the path (index of the middle child node)
     ///     * `child_mem_usage` – memory usage of the middle child
     ///     * `left_mem_usage` – total memory usage of left siblings of the middle child
     /// Returns `None` if the end of the searched is reached.
-    fn next_step(&self) -> Option<(u8, u64, u64)> {
+    fn next_step(&self) -> SplitResult<Option<(u8, u64, u64)>> {
         // Stop when a leaf is reached in the accounts subtree
         if !self.subtree_stages[0].can_descend() {
             tracing::debug!(target = "memtrie", "leaf reached in accounts subtree");
-            return None;
+            return Ok(None);
         }
 
-        let current_nodes_memory = self.current_nodes_mem_usage();
-        let threshold = (self.total_memory() / 2)
-            .checked_sub(self.left_memory)?
-            .checked_sub(current_nodes_memory)?;
-
-        let children_mem_usage = self.aggregate_children_mem_usage();
+        let current_nodes_memory = self.current_nodes_mem_usage()?;
+        if self.left_memory + current_nodes_memory > self.total_memory() / 2 {
+            return Ok(None);
+        }
+        let threshold = (self.total_memory() / 2) - self.left_memory - current_nodes_memory;
+        let children_mem_usage = self.aggregate_children_mem_usage()?;
 
         // Find the middle child
         tracing::debug!(target = "memtrie", ?children_mem_usage, %threshold, "finding middle child");
-        let (middle_child, left_mem_usage) = find_middle_child(&children_mem_usage, threshold)?;
+        let Some((middle_child, left_mem_usage)) =
+            find_middle_child(&children_mem_usage, threshold)
+        else {
+            return Ok(None);
+        };
         let child_mem_usage = children_mem_usage[middle_child];
         tracing::debug!(target = "memtrie", %middle_child, %child_mem_usage, %left_mem_usage, "middle child found");
 
         // Stop if the further path does not exist in the accounts subtree
         let middle_child = middle_child as u8;
-        self.subtree_stages[0].can_descend_nibble(middle_child).then_some((
+        Ok(self.subtree_stages[0].can_descend_nibble(middle_child).then_some((
```

Comment:
nit: How about the following?
Also I slightly dislike the 0 index as the account subtree, maybe use a constant for it? 
```
        let can_descend = self.subtree_stages[0].can_descend_nibble(middle_child);
        if !can_descend {
            return Ok(None);
        }
        Ok(Some((middle_child, child_mem_usage, left_mem_usage)))
```